### DISCUSSION TITLE
Wiring NFC reader

### DISCUSSION THREAD

**Discussion by thorsten-gehrig** [rednblkx/HomeKey-ESP32#13] [General]
Hi
earlier you had in your source defined all the NFC-reader pins:
`// // If using the breakout with SPI, define the pins for SPI communication.
#define PN532_SCK (D5)
#define PN532_MOSI (D7)
#define PN532_SS (D8)
#define PN532_MISO (D6)
#define RESET_PIN (D3)

PN532_SPI pn532spi(SPI, PN532_SS);
PN532 nfc(pn532spi);`

In the refactored code you have only the pin 5 defined - and i cannot find other references.
`#define PN532_SS (5)
PN532_SPI pn532spi(SPI, PN532_SS);`

So... stupid question: which pins/wiring is finally required to have this up and running (beside obvious VCC and GND).
I don't find definitions of SCK, MOSI, MISO and RESET in the code.
Is it only PIN5? (IO5 // D8)?
Trying to understand the SPI protocol it seems MISO/MOSI/SS (or "[new](https://learn.sparkfun.com/tutorials/serial-peripheral-interface-spi/all#:~:text=Serial%20Peripheral%20Interface%20(SPI)%20is,you%20wish%20to%20talk%20to.)" POCI/PICO/CS) are mandatory for SPI... so what did i miss?

Best
Thorsten

**Comment by thorsten-gehrig** [comment]
I guess i found my answer: according to this schematics the MISO/MOSI/SCK/CS0 are defined by default...
https://www.fambach.net/d1-mini-esp8266-modul-2-2/

Side question: would it be possible to use I2C instead of SPI? Should work as well, isnt?

**Comment by jklap** [comment]
```
#include <PN532_I2C.h>
PN532_I2C pn532i2c(Wire);
PN532 nfc(pn532i2c);
```

You could try swapping out the existing nfc/PN532 setup at the top of `main.cpp`

The use of `Wire` above defines/uses bus 0. Or use `Wire1` which uses bus 1. Bus 0 uses GPIO 21/22 (usual pins).

Also beware that there seem to be lots of issues with PN532 modules (especially low-quality/cheap/counterfeit) and use of I2C.

**Comment by rednblkx** [comment]
Hi Thorsten,

Glad you were able to find the answer and yes, the ESP32 like Arduino boards it has denominated pins for SPI and such.

Regarding I2C, i tried it, couldn’t make it work although didn’t really put a lot of effort into it since SPI already works but the PN532 library has it implemented so i suppose it should work

Also, thanks for pointing this out, I’ll be updating the README to including the wiring which should’ve been there from the beginning 🤡
